samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs

```diff
@@ -110,19 +109,56 @@ public async Task<ActionResult> SendNotificationToOrganisation(BroadcastInfo tas
 
-                Parallel.ForEach (usersList, async users =>
-                {
-                    var installedApp = await graphClient.Users[users.Id].Teamwork.InstalledApps
+                await UtilityHelper.ForEachAsync(usersList, 5, async users => {
+                    try
+                    {
+                        var installedApp = await graphClient.Users[users.Id].Teamwork.InstalledApps
                                              .Request()
                                              .Expand("teamsApp")
                                              .GetAsync();
 
-                    var response = new HttpResponseMessage();
-                    var installationId = installedApp.Where(id => id.TeamsApp.DisplayName == "Activity feed broadcast").Select(x => x.TeamsApp.Id);
-                    var client = new HttpClient();
-                    var url = "https://teams.microsoft.com/l/entity/" + appId.ToList()[0] + "/broadcast?context={\"subEntityId\":\"" + taskInfo.taskId + "\"}";
+                        var response = new HttpResponseMessage();
+                        var installationId = installedApp.Where(id => id.TeamsApp.DisplayName == _configuration["AppName"]).Select(x => x.TeamsApp.Id);
+
+                        if (installationId.ToList().Count == 0)
+                        {
+                            var userScopeTeamsAppInstallation = new UserScopeTeamsAppInstallation
+                            {
+                                AdditionalData = new Dictionary<string, object>()
+                            {
+                                {"teamsApp@odata.bind", "https://graph.microsoft.com/v1.0/appCatalogs/teamsApps/"+appId.ToList()[0]}
+                            }
+                            };
+
+                            await graphClient.Users[users.Id].Teamwork.InstalledApps
+                                                .Request()
+                                                .AddAsync(userScopeTeamsAppInstallation);
+                        }
+                        counter++;
+                        recipientsList.Add(new Dictionary<string, string>
+                        {
+                            {"@odata.type", "microsoft.graph.aadUserNotificationRecipient" },
+                            {"userId", users.Id}
+                        });
+                    }
+                    catch (Exception ex)
+                    {
+                        Console.WriteLine("Installation failed for " + users.DisplayName);
+                    }
+                });
+
+                var url = "https://teams.microsoft.com/l/entity/" + appId.ToList()[0] + "/broadcast?context={\"subEntityId\":\"" + taskInfo.taskId + "\"}";
+                var client = new HttpClient();
+
+                var recipientsChunks = UtilityHelper.SplitIntoChunks(recipientsList, 85);
+
+                foreach (var recipientChunk in recipientsChunks)
```

Comment:
Rename it to recipient

Reply:
Updated to recipientList as its not individual entity.